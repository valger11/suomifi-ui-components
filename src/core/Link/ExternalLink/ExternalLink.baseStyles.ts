import { css } from 'styled-components';
import { suomifiTheme } from '../../theme';
import { baseStyles } from '../BaseLink/BaseLink.baseStyles';

export const ExternalLinkStyles = css`
  ${baseStyles}
  & .fi-link_icon {
    padding-left: ${suomifiTheme.spacing.insetXs};
    font-size: 16px;
    box-sizing: content-box;
  }
`;
