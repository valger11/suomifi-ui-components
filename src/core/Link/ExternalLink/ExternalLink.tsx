import React, { Component } from 'react';
import { default as styled } from 'styled-components';
import { logger } from '../../../utils/log';
import classnames from 'classnames';
import { SuomifiThemeProp, SuomifiThemeConsumer } from '../../theme';
import { Icon } from '../../Icon/Icon';
import { VisuallyHidden } from '../../VisuallyHidden/VisuallyHidden';
import { ExternalLinkStyles } from './ExternalLink.baseStyles';
import { HtmlA } from '../../../reset';
import { BaseLinkProps, baseClassName } from '../BaseLink/BaseLink';

const iconClassName = 'fi-link_icon';
const externalClassName = 'fi-link--external';

export interface ExternalLinkProps extends BaseLinkProps {
  /** Translated explanation of 'opens to a new window' */
  labelNewWindow: string;
  /** Hide the icon */
  hideIcon?: boolean;
  /** Open to a new window
   * @default true
   */
  toNewWindow?: boolean;
}

class BaseExternalLink extends Component<ExternalLinkProps> {
  render() {
    const {
      asProp,
      children,
      className,
      toNewWindow = true,
      labelNewWindow,
      hideIcon,
      ...passProps
    } = this.props;
    return (
      <HtmlA
        {...passProps}
        className={classnames(baseClassName, className, externalClassName)}
        target={!!toNewWindow ? '_blank' : undefined}
        rel={!!toNewWindow ? 'noopener' : undefined}
        as={asProp}
      >
        {children}
        <VisuallyHidden>{labelNewWindow}</VisuallyHidden>
        {!hideIcon && (
          <Icon
            icon="linkExternal"
            className={iconClassName}
            color="currentColor"
          />
        )}
      </HtmlA>
    );
  }
}

const StyledExternalLink = styled(
  (props: ExternalLinkProps & SuomifiThemeProp) => {
    const { theme, ...passProps } = props;
    return <BaseExternalLink {...passProps} />;
  },
)`
  ${({ theme }) => ExternalLinkStyles(theme)};
`;

/**
 * <i class="semantics" />
 * Used for adding a external site link
 */
export class ExternalLink extends Component<ExternalLinkProps> {
  render() {
    const { labelNewWindow, ...passProps } = this.props;
    if (!labelNewWindow) {
      logger.warn(
        'External link needs a translated description of link opening to a new window',
      );
    }
    return (
      <SuomifiThemeConsumer>
        {({ suomifiTheme }) => (
          <StyledExternalLink
            theme={suomifiTheme}
            labelNewWindow={labelNewWindow}
            {...passProps}
          />
        )}
      </SuomifiThemeConsumer>
    );
  }
}
