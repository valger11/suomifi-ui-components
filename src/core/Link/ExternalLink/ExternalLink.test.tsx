import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../../utils/test/axe';

import { ExternalLink } from './ExternalLink';

const TestExternalLink = (
  <ExternalLink
    href="/"
    data-testid="test-link"
    labelNewWindow="Opens in a new window"
  >
    Hey this is test
  </ExternalLink>
);

test('calling render with the same component on the same container does not remount', () => {
  const LinkRendered = render(TestExternalLink);
  const { getByTestId, container } = LinkRendered;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('test-link').textContent).toBe(
    'Hey this is testOpens in a new window',
  );
});

test('should not have basic accessibility issues', axeTest(TestExternalLink));
