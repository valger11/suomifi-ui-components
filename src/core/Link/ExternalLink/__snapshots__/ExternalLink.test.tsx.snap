// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c4 {
  display: inline-block;
  vertical-align: baseline;
}

.c4 .fi-icon-base-fill {
  fill: currentColor;
}

.c4 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c4.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c4.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c3 {
  position: absolute;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c0:hover,
.c0:active,
.c0:focus,
.c0:focus-within {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.c1 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:hover,
.c1:active,
.c1:focus,
.c1:focus-within {
  color: hsl(212,63%,45%);
}

.c1:focus,
.c1:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:focus {
  outline: 0;
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
}

.c1:hover,
.c1:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:visited {
  color: hsl(284,36%,45%);
}

.c1 .fi-link_icon {
  padding-left: 4px;
  font-size: 16px;
  box-sizing: content-box;
}

<a
  class="c0 fi-link c1 fi-link--external"
  data-testid="test-link"
  href="/"
  rel="noopener"
  target="_blank"
>
  Hey this is test
  <span
    class="c2 c3 fi-visually-hidden"
  >
    Opens in a new window
  </span>
  <svg
    aria-hidden="true"
    class="fi-icon c4 fi-link_icon"
    focusable="false"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      class="fi-icon-base-fill"
      d="M12 4a1 1 0 010 2H2v16h16V12a1 1 0 012 0v10c0 1.103-.897 2-2 2H2c-1.103 0-2-.897-2-2V6c0-1.103.897-2 2-2h10zm11-4a1 1 0 011 1v6.657a1 1 0 01-2 0l-.001-4.242-9.792 9.792a.997.997 0 01-1.414 0 .999.999 0 010-1.414L20.586 2h-4.243a1 1 0 010-2H23z"
      fill="#292929"
      fill-rule="evenodd"
    />
  </svg>
</a>
`;
